test/src/loaders/workspace_uploader/workspace_uploader_test.py

```diff
@@ -1,5 +1,302 @@
+        conf, wait_to_upload_assemblies
+    )
+
+    assert sorted(os.listdir(data_dir)) == sorted(ASSEMBLY_NAMES)
+    for assembly_name, src_file in zip(ASSEMBLY_NAMES, params.target_files):
+        assert os.path.samefile(src_file, os.path.join(data_dir, assembly_name))
+
+
+def test_post_process(setup_and_teardown):
+    params = setup_and_teardown
+    upload_dir = Path(params.tmp_dir) / "upload_dir"
+    upload_dir.mkdir()
+    output_dir = Path(params.tmp_dir) / "output_dir"
+    output_dir.mkdir()
+
+    upa = "12345_58_1"
+    upload_env_key = "CI"
+    workspace_id = 88888
+    host_assembly_dir = params.assembly_dirs[0]
+    assembly_name = ASSEMBLY_NAMES[0]
+    src_file = params.target_files[0]
+
+    workspace_uploader._post_process(
+        upload_env_key,
+        workspace_id,
+        host_assembly_dir,
+        assembly_name,
+        upload_dir,
+        output_dir,
+        upa,
+    )
+
+    data, uploaded = workspace_uploader._read_upload_status_yaml_file(
+        upload_env_key, workspace_id, host_assembly_dir, assembly_name
+    )
+    expected_data = {
+        upload_env_key: {workspace_id: {"file_name": assembly_name, "upa": upa}}
+    }
+
+    dest_file = os.path.join(os.path.join(output_dir, upa), f"{upa}.fna.gz")
+
+    assert uploaded
+    assert expected_data == data
+    # check softlink
+    assert os.readlink(os.path.join(upload_dir, upa)) == os.path.join(output_dir, upa)
+    # check hardlink
+    assert os.path.samefile(src_file, dest_file)
+
+
+def test_upload_assembly_to_workspace(setup_and_teardown):
+    params = setup_and_teardown
+    workspace_id = 12345
+    assembly_name = ASSEMBLY_NAMES[0]
+    file_path = "/path/to/file/in/AssembilyUtil"
+
+    conf = create_autospec(Conf)
+    conf.asu = create_autospec(AssemblyUtil)
+    conf.asu.save_assembly_from_fasta2.return_value = {"upa": "12345/58/1"}
+    upa = workspace_uploader._upload_assembly_to_workspace(
+        conf, workspace_id, file_path, assembly_name
+    )
+
+    assert upa == "12345_58_1"
+    assert conf.asu.save_assembly_from_fasta2.call_count == 1
+
+
+def test_assembly_files_in_parallel(setup_and_teardown):
+    params = setup_and_teardown
+    upload_dir = Path(params.tmp_dir) / "upload_dir"
+    upload_dir.mkdir()
+
+    upload_env_key = "CI"
+    workspace_id = 12345
+    upa = "12345_58_1"
+    assembly_dirs = params.assembly_dirs
+
+    wait_to_upload_assemblies = {
+        assembly_name: assembly_dir
+        for assembly_name, assembly_dir in zip(ASSEMBLY_NAMES, assembly_dirs)
+    }
+
+    conf = Mock()
+    conf.workers = 5
+    conf.input_queue = Queue()
+    conf.output_queue = Queue()
+
+    # an uploaded successful
+    conf.output_queue.put((ASSEMBLY_NAMES[0], upa))
+    # an upload failed
+    conf.output_queue.put((ASSEMBLY_NAMES[1], None))
+
+    failed_names = workspace_uploader._upload_assembly_files_in_parallel(
+        conf, upload_env_key, workspace_id, upload_dir, wait_to_upload_assemblies
+    )
+
+    (
```

Comment:
What I would do here, rather than testing each item of the queue tuple once by one, is make expected tuples and then 

```
assert conf.input_queue.get() == expected_tuple1
assert conf.input_queue.get() == expected_tuple2
```

pytest is good about providing error messages showing where the differences in data structures exist

Reply:
Updated!

Reply:
:+1: